gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/computer/bulkdumping/BulkExportVertexProgram.java

```diff
@@ -0,0 +1,189 @@
+
+    private Configuration configuration;
+    private Map<String, String> properties;
+    private List<String> sortedProperties;
+    private Set<VertexComputeKey> vertexComputeKeys;
+
+    private BulkExportVertexProgram() {
+    }
+
+    @Override
+    public void loadState(final Graph graph, final Configuration config) {
+        configuration = new BaseConfiguration();
+        if (config != null) {
+            ConfigurationUtils.copy(config, configuration);
+        }
+        properties = new HashMap<>();
+        sortedProperties = new ArrayList<>();
+        for (final String tuple : configuration.getString(BULK_EXPORT_PROPERTIES, "").split("\1")) {
+            final String[] parts = tuple.split("\2", -1);
+            properties.put(parts[0], parts[1]);
+            sortedProperties.add(parts[0]);
+        }
+        vertexComputeKeys = Collections.singleton(VertexComputeKey.of(BULK_EXPORT_PROPERTIES, false));
+    }
+
+    @Override
+    public void storeState(final Configuration config) {
+        VertexProgram.super.storeState(config);
+        if (configuration != null) {
+            ConfigurationUtils.copy(configuration, config);
+        }
+    }
+
+    @Override
+    public void setup(final Memory memory) {
+    }
+
+    @Override
+    public void execute(final Vertex sourceVertex, final Messenger<Tuple> messenger, final Memory memory) {
+        final VertexProperty<TraverserSet> haltedTraversers = sourceVertex.property(TraversalVertexProgram.HALTED_TRAVERSERS);
+        haltedTraversers.ifPresent(traverserSet -> {
+            final List<List<String>> rows = new ArrayList<>();
+            for (final Traverser t : (Iterable<Traverser>) traverserSet) {
+                final List<String> columns = new ArrayList<>();
+                final Path path = t.path();
+                final Iterable<String> keys = properties.isEmpty()
+                        ? t.path().labels().stream().flatMap(Collection::stream).sorted().collect(Collectors.toSet())
+                        : sortedProperties;
+                for (final String key : keys) {
+                    final String format = properties.getOrDefault(key, "");
+                    final Object value = path.get(key);
+                    columns.add("".equals(format) ? value.toString() : String.format(format, value));
+                }
+                rows.add(columns);
+            }
+            sourceVertex.property(BULK_EXPORT_PROPERTIES, rows);
+        });
+    }
+
+    @Override
+    public boolean terminate(final Memory memory) {
+        return properties == null || properties.isEmpty() || !memory.isInitialIteration();
+    }
+
+    @Override
+    public Set<MessageScope> getMessageScopes(final Memory memory) {
+        return Collections.emptySet();
+    }
+
+    @SuppressWarnings({"CloneDoesntDeclareCloneNotSupportedException", "CloneDoesntCallSuperClone"})
+    @Override
+    public VertexProgram<Tuple> clone() {
```

Comment:
Extend `StaticVertexProgram<Tuple>` instead of implementing `VertexProgram<Tuple>` and you can get rid of this `clone()` method (and others I believe?).